### PR TITLE
(PUP-10255) Follow symlinks in environment path

### PR DESCRIPTION
**Pull request by mwaggett**
This commit adds a `real_path` method that will follow symlinks in the given
directory, which is then used when getting the environment.conf and generating
metadata for a given environment. This enables us to have multiple code
dirs for a given environment, with the set `environmentpath` pointing to the
most recent version, which enables lockless deployments.

### CODE REVIEW

File: lib/puppet/file_system/memory_impl.rb [hunk 1]
```diff
@@ -23,6 +23,10 @@ def executable?(path)
     path.executable?
   end
 
+  def symlink(path)
```
Comment (mwaggett) [inline lib/puppet/file_system/memory_impl.rb]: I hoped this would fix the failing tests, but doesn't look like it.. 🤔

Comment (mwaggett) [reply lib/puppet/file_system/memory_impl.rb]: oh I'm an idiot

Comment (mwaggett) [reply lib/puppet/file_system/memory_impl.rb]: 🤞 putting the question mark in there fixes it.. 🤞

**Comment by mwaggett** [comment]
jenkins please test this with servertests

**Comment by mwaggett** [comment]
also tested this locally by
- setting up a test file:
```
molly.waggett@molly:~/.puppetlabs $ cat opt/server/data/testing/environments/production/manifests/site.pp
notify { 'hello': }
```
- creating a symlink where the environment path should be:
```
molly.waggett@molly:~/.puppetlabs $ ls -l etc/code/environments/production
lrwxr-xr-x  1 molly.waggett  staff  81 Feb  3 14:10 etc/code/environments/production -> /Users/molly.waggett/.puppetlabs/opt/server/data/testing/environments/production/
```
- updating the puppet submodule in puppetserver to this PR
- running puppetserver from the repl
- doing a puppet run:
```
molly.waggett@molly:~/puppet-projects/puppet(ticket/master/pup-10255-symlinked-environment-paths👌 ) $ bundle exec puppet agent -t --server localhost
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Caching catalog for localhost
Info: Applying configuration version '1580873427'
Notice: hello
Notice: /Stage[main]/Main/Notify[hello]/message: defined 'message' as 'hello'
Notice: Applied catalog in 0.01 seconds
```

**Comment by Magisus** [comment]
It might be nice to have some spec tests here, maybe in https://github.com/puppetlabs/puppet/blob/master/spec/unit/environments_spec.rb?

**Comment by mwaggett** [comment]
so yeah, I was thinking about what tests might make sense.. 
I could write a test for `real_path`, and maybe some of the other helper methods, but that sort of feels like an implementation detail to me? And then what would the test look like? I feel like it would just be stubbing everything out such that we'd really just be testing that `Puppet::FileSystem.readlink` works correctly (which I believe we already do).

I agree that it feels weird to not include tests with this.. Did you have something in mind that seemed test-worthy?

**Comment by mwaggett** [comment]
okay, I'm writing tests in https://github.com/puppetlabs/puppet/blob/master/spec/unit/environments_spec.rb, didn't realize we mocked a whole directory structure in there, so shouldn't be too hard to check for symlinks.

File: lib/puppet/environments.rb [hunk 2]
```diff
@@ -230,20 +233,22 @@ def create_environment(name)
...
-         Puppet::Node::Environment.valid_name?(name)
+    def validated_directory(envdir)
+      env_name = Puppet::FileSystem.basename_string(envdir)
+      envdir = Puppet::Environments::Directories.real_path(envdir)
+      return nil unless Puppet::FileSystem.directory?(envdir) && Puppet::Node::Environment.valid_name?(env_name)
+      return envdir
     end
 
-    def valid_directories
+    def valid_directories_by_env
```
Comment (mwaggett) [inline lib/puppet/environments.rb]: Nothing actually uses the actual directories from this, but I kept 'em since the previous version of the method returned those. Might clean up later though..

**Comment by Magisus** [comment]
In your commit message, I think it should `environment.conf` instead of `puppet.conf`

**Comment by mwaggett** [comment]
oh ya, will update that

**Comment by mwaggett** [comment]
okay, fixed the commit message

**Comment by Magisus** [comment]
jenkins please test this with servertests

File: lib/puppet/environments.rb [hunk 3]
```diff
@@ -230,20 +233,22 @@ def create_environment(name)
...
-         Puppet::Node::Environment.valid_name?(name)
+    def validated_directory(envdir)
+      env_name = Puppet::FileSystem.basename_string(envdir)
+      envdir = Puppet::Environments::Directories.real_path(envdir)
+      return nil unless Puppet::FileSystem.directory?(envdir) && Puppet::Node::Environment.valid_name?(env_name)
+      return envdir
     end
 
-    def valid_directories
+    def valid_environments
```
Comment (justinstoller) [inline lib/puppet/environments.rb]: why the name change?

File: lib/puppet/environments.rb [hunk 4]
```diff
@@ -230,20 +233,22 @@ def create_environment(name)
...
+      envs = []
       if Puppet::FileSystem.directory?(@environment_dir)
-        Puppet::FileSystem.children(@environment_dir).select do |child|
-          valid_directory?(child)
+        Puppet::FileSystem.children(@environment_dir).each do |child|
```
Comment (justinstoller) [inline lib/puppet/environments.rb]: why go from `select` to appending to an array here?

File: lib/puppet/environments.rb [hunk 5]
```diff
@@ -230,20 +233,22 @@ def create_environment(name)
...
     end
 
-    def valid_directory?(envdir)
-      name = Puppet::FileSystem.basename_string(envdir)
-      Puppet::FileSystem.directory?(envdir) &&
-         Puppet::Node::Environment.valid_name?(name)
+    def validated_directory(envdir)
+      env_name = Puppet::FileSystem.basename_string(envdir)
+      envdir = Puppet::Environments::Directories.real_path(envdir)
+      return nil unless Puppet::FileSystem.directory?(envdir) && Puppet::Node::Environment.valid_name?(env_name)
```
Comment (justinstoller) [inline lib/puppet/environments.rb]: I don't know if this is just a me thing, but I much prefer a conditional the returns false/nil explicitly rather than two sequential `return`s
eg
```rb
if directory?(envdir) && valid_name?(name)
  name
else
  nil
end
```

File: lib/puppet/environments.rb (hunk 3 above)
Comment (mwaggett) [reply lib/puppet/environments.rb]: eh, I had refactored this a bunch and `valid_directories` didn't seem as accurate anymore, though now I see that I landed more closely to the original than I had in some in-between iterations. We're returning a list of environment names, so this seemed clearer to me, but I can put it back if you'd prefer.

File: lib/puppet/environments.rb (hunk 4 above)
Comment (mwaggett) [reply lib/puppet/environments.rb]: another remnant of my in-between refactor. but I guess this now returns just env names, rather than whole directories?

File: lib/puppet/environments.rb (hunk 5 above)
Comment (mwaggett) [reply lib/puppet/environments.rb]: heh, yeah, this is something I picked up from Eric G on RE. I think the idea is that the "happy-path" case shouldn't be in a conditional?

File: lib/puppet/environments.rb (hunk 3 above)
Comment (justinstoller) [reply lib/puppet/environments.rb]: I expected this to be a list of `Environment`s, not to be super pedantic but I think we should name things to keep "whole paths", "directory names", and actual Environments separate.

File: lib/puppet/environments.rb (hunk 4 above)
Comment (justinstoller) [reply lib/puppet/environments.rb]: I think if you change it back to `select` and `env_name if validated_directory(child)` will return nil if it doesn't match and select won't include it in the output (and if everything is removed it will return an empty array. Removing the need for `envs`.

File: lib/puppet/environments.rb (hunk 3 above)
Comment (mwaggett) [reply lib/puppet/environments.rb]: okay, that's fair. I guess when I think of a "directory" I imagine a path? What do you think of `valid_environment_directories`? or `valid_environment_directory_names`? Are those too wordy? Maybe I'll just put it back to how it was..?

File: lib/puppet/environments.rb (hunk 4 above)
Comment (mwaggett) [reply lib/puppet/environments.rb]: yeah, I can go back to `select`

File: lib/puppet/environments.rb (hunk 5 above)
Comment (justinstoller) [reply lib/puppet/environments.rb]: I think I disagree with that, though I'm not sure exactly if I follow the rationale.

I'd argue that save for a quick input sanity check at the top of a method early return should be avoided (and even those should be minimized).

I'm aware that this method is also too small to care a lot about, but now I'm curious what other's opinions are?

/cc @tvpartytonight @Magisus @pcarlisle ?

File: lib/puppet/environments.rb (hunk 3 above)
Comment (justinstoller) [reply lib/puppet/environments.rb]: I don't think they're too wordy, `valid_environment_names` would work too?

File: lib/puppet/environments.rb (hunk 4 above)
Comment (mwaggett) [reply lib/puppet/environments.rb]: hmm actually I don't know if that will work because I'm still not trying to return the actual `child`s.. might need to `map`?

File: lib/puppet/environments.rb (hunk 3 above)
Comment (mwaggett) [reply lib/puppet/environments.rb]: I like `valid_environment_names` 😊

File: lib/puppet/environments.rb (hunk 5 above)
Comment (Magisus) [reply lib/puppet/environments.rb]: The sentiment I've heard is to avoid large indented blocks, which goes along with what Justin described: if you need to bail before doing any work because something is wrong, an early return is good for that, instead of having large chunks wrapped under `if` and `else`. In this case, since the cases are small and clearly binary, I think I prefer Justin's. Another plus, it avoids the `unless` ;)

Comment (mwaggett) [reply lib/puppet/environments.rb]: What's the rationale for avoiding early `return`s?

Comment (justinstoller) [reply lib/puppet/environments.rb]: It can be confusing. Again this method isn't large enough to really suffer from it, but returns in the middle of a method body can be easily missed and when you have sequential returns based on different conditionals I think they can sometimes interact in non-obvious ways. In that case I think it's usually clearer to put all the conditionals in a single control structure (if/elsif/else block) to be explicit about how they interact.

File: lib/puppet/environments.rb (hunk 4 above)
Comment (justinstoller) [reply lib/puppet/environments.rb]: oh good point

File: lib/puppet/environments.rb (hunk 3 above)
Comment (mwaggett) [reply lib/puppet/environments.rb]: updated

File: lib/puppet/environments.rb (hunk 4 above)
Comment (mwaggett) [reply lib/puppet/environments.rb]: updated

File: lib/puppet/environments.rb (hunk 5 above)
Comment (mwaggett) [reply lib/puppet/environments.rb]: mmkay, I suppose that's fair. I updated it.

File: lib/puppet/environments.rb [hunk 6]
```diff
@@ -186,34 +186,37 @@ def self.from_path(path, global_module_path)
...
     # @!macro loader_get_conf
     def get_conf(name)
-      envdir = File.join(@environment_dir, name.to_s)
-      if valid_directory?(envdir)
-        return Puppet::Settings::EnvironmentConf.load_from(envdir, @global_module_path)
-      end
-      nil
+      envdir = validated_directory(File.join(@environment_dir, name.to_s))
+      return nil unless envdir
+      return Puppet::Settings::EnvironmentConf.load_from(envdir, @global_module_path)
```
Comment (justinstoller) [inline lib/puppet/environments.rb]: fwiw, this could be `if envdir` ... I think?

Comment (mwaggett) [reply lib/puppet/environments.rb]: What could? Are you saying I could reverse the order of these `return`s? Or are you saying I should change this to an `if`/`else` like the other one?
